x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportCloseJobAction.java

```diff
@@ -116,30 +116,39 @@ protected void doExecute(Task task, CloseJobAction.Request request, ActionListen
                                         Set<String> executorNodes = new HashSet<>();
                                         PersistentTasksCustomMetaData tasks = state.metaData().custom(PersistentTasksCustomMetaData.TYPE);
                                         for (String resolvedJobId : request.getOpenJobIds()) {
                                             PersistentTasksCustomMetaData.PersistentTask<?> jobTask =
-                                                    MlTasks.getJobTask(resolvedJobId, tasks);
-
-                                            if (jobTask == null || jobTask.isAssigned() == false) {
-                                                String message = "Cannot close job [" + resolvedJobId + "] because the job does not have "
-                                                        + "an assigned node. Use force close to close the job";
-                                                listener.onFailure(ExceptionsHelper.conflictStatusException(message));
-                                                return;
-                                            } else {
+                                                MlTasks.getJobTask(resolvedJobId, tasks);
+
+                                            if (jobTask == null) {
+                                                // This should not happen, because openJobIds was
+                                                // derived from the same tasks metadata as jobTask
+                                                String msg = "Requested job [" + resolvedJobId
+                                                    + "] be stopped, but job's task could not be found.";
+                                                assert jobTask != null : msg;
+                                                logger.error(msg);
+                                            } else if (jobTask.isAssigned()) {
                                                 executorNodes.add(jobTask.getExecutorNode());
+                                            } else {
+                                                // This is the easy case - the job is not currently assigned to a node, so can
+                                                // be gracefully stopped simply by removing its persistent task.  (Usually a
+                                                // graceful stop cannot be achieved by simply removing the persistent task, but
+                                                // if the job has no running code then graceful/forceful are basically the same.)
+                                                // The listener here can be a no-op, as waitForJobClosed() already waits for
+                                                // these persistent tasks to disappear.
+                                                persistentTasksService.sendRemoveRequest(jobTask.getId(),
+                                                    ActionListener.wrap(r -> {}, e -> {}));
```

Comment:
Should we not handle an error in case the removal of the persistent task fails?

Reply:
I'll add logging of the error at this point in the code.  To do anything more would mean completely refactoring the way the class waits for the per-job requests.

The overall request will time out instead of reporting success if removing the persistent task for an unassigned job fails.  This is similar to what happens if removing the persistent task for an assigned job fails after we've done all the other steps.  Effectively the change is making treatment of assigned and unassigned jobs more similar.